rfcs/0000-cosign-integration.md

```diff
@@ -0,0 +1,162 @@
+by `cosign` signing should be annotated with `kpack.io/cosign-credentials`.
+Optionally, the user can also annotate the secrets with the following
+information:
+- `kpack.io/cosign.repository`: a separate registry URL to upload the generated
+signatures, instead of pushing the signatures collocated with the images. Adding
+this annotation should have the same effect as setting the `COSIGN_REPOSITORY`
+variable. The user must provide credentials for `kpack` to authenticate with the
+specified registry. In the case where this repository is inaccessible, the build
+should fail and output an error message to allow the operator to troubleshoot.
+- `kpack.io/cosign.docker-media-types`: a flag indicating whether legacy media
+types should be used instead of OCI media types. Adding this annotation should
+have the same effect as setting the `COSIGN_DOCKER_MEDIA_TYPES` variable.
+
+Secret format example:
+```yaml
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: cosign-key
+  annotations:
+    kpack.io/cosign-credentials: ""
+    # Optional
+    kpack.io/cosign.repository: "<repository-url>"
+    # Optional
+    kpack.io/cosign.docker-media-types: "1"
+type: Opaque
+stringData:
+  cosign.key: <cosign-private-key>
+  cosign.pub: <cosign-public-key>
+  cosign.password: <cosign-private-key-passphrase>
+```
+
+### Configure custom static annotations
+
+Create a new optional configuration in the `Image` resource that allow users to
+parameterize optional annotations to add to the generated signatures. These
+annotations will be static and configured as a list of key-value pairs:
+```yaml
+cosign:
+  annotations:
+  - name: "key1"
+    value: "value1"
+  - name: "key2"
+    value: "value2"
+```
+
+### Authentication to a container registry
+
+`cosign` uses the `DefaultKeychain` from
+[`go-containerregistry`](https://github.com/google/go-containerregistry/blob/main/pkg/authn/README.md#tldr-for-consumers-of-this-package)
+to authenticate the command-line interface using the machine's Docker CLI
+credentials. For `kpack`, it may be possible to use the same authentication
+mechanism that is in-place for uploading an image.
+
+## Complexity
+
+Integration with `cosign` is estimated to be of medium complexity. Since
+`cosign` is also written in Go, it is possible to import code from the tool
+directly within `kpack`, which can help make the integration easier.
+
+## Prior art
+
+- [Issue](https://github.com/pivotal/kpack/issues/684) filed on May 4th.
+- [Use of `cosign` in Kubernetes](https://github.com/kubernetes/release/pull/2016)
+  for validation of distroless images.
+- [Use of `cosign` in Connaisseur](https://github.com/sse-secure-systems/connaisseur/pull/107)
+  for policy enforcement in container images.
+- [Notary v1 implementation pull request](https://github.com/pivotal/kpack/pull/541).
+
+## Alternatives
+
+- Use a CI/CD step to sign images using `cosign` instead of implementing it as
```

Comment:
Could we mention the alternative of configuring the cosign configuration on the image spec?

Reply:
Sure, I'll add it here!